### PR TITLE
Feature/sound

### PR DESCRIPTION
**Pull request by JucanAndreiDaniel**
Implemented multiple features used for Sound playback and decoding

### CODE REVIEW

**Comment by JucanAndreiDaniel** [comment]
Need to add testing for audio manipulation but will do in another PR

File: PxCs/Interface/PxSound.cs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+            try
+            {
+                ulong size = PxBuffer.pxBufferSize(wavSound);
+
+                var arrayPtr = IntPtr.Zero;
+                arrayPtr = PxBuffer.pxBufferArray(wavSound);
+                byte[] wavFile = new byte[size];
+
+                Marshal.Copy(arrayPtr, wavFile, 0, (int)size);
```
Comment (JaXt0r) [inline PxCs/Interface/PxSound.cs:41]: Please use IntPtr.MarshalAsArray<byte>(lenght) as it's proven and used already.

File: PxCs/Interface/PxSound.cs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+using PxCs.Extensions;
+
+
+namespace PxCs.Interface
+{
+    public class PxSound
+    {
+        private const string DLLNAME = PxPhoenix.DLLNAME;
+
+        public static byte[] GetSoundByteArrayFromVDF(IntPtr vdfPtr, string name)
```
Comment (JaXt0r) [inline PxCs/Interface/PxSound.cs:13]: This is an uppercased method. Can you please return a PxData object. Look inside other Interface classes. This one should follow similarly please.

E.g. this is how Mrm is fetched from PxCs: https://github.com/GothicVRProject/GothicVR/blob/feature/audio-implementation/Assets/GothicVR/Scripts/Caches/AssetCache.cs#L119

File: PxCs/Interface/PxSound.cs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                var arrayPtr = IntPtr.Zero;
+                arrayPtr = PxBuffer.pxBufferArray(wavSound);
+                byte[] wavFile = new byte[size];
+
+                Marshal.Copy(arrayPtr, wavFile, 0, (int)size);
+                return wavFile;
+            }
+            finally
+            {
+                PxBuffer.pxBufferDestroy(wavSound);
```
Comment (JaXt0r) [inline PxCs/Interface/PxSound.cs:46]: Great, that you thought about it! #NoPowerToMemoryLeaks!

File: PxCs/Interface/PxSound.cs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            }
+        }
+
+        public static float[] GetSoundFloatArrayFromVDF(IntPtr vdfPtr, string name)
```
Comment (JaXt0r) [inline PxCs/Interface/PxSound.cs:50]: Same as above method: How about returning a Px...Data Object?

File: PxCs/Interface/PxSound.cs [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+                default:
+                    throw new Exception(bitDepth + " bit depth is not supported.");
+            }
+
+            return data;
+        }
+
+        #region wav file bytes to Unity AudioClip conversion methods
+
+        private static float[] Convert8BitByteArrayToFloatArray(byte[] source, int headerOffset, int dataSize)
```
Comment (JaXt0r) [inline PxCs/Interface/PxSound.cs:102]: How about having only one method which takes another parameter called from/to? With this you would decide from outside what byte[] type you send and want to retreive? Similar to this function: https://github.com/GothicKit/phoenix-csharp-interface/blob/main/PxCs/Interface/PxTexture.cs#L41

File: PxCs/Extensions/SoundExtension.cs [hunk 6]
```diff
@@ -0,0 +1,302 @@
+using System;
+using System.Text;
+using System.IO;
+using System.Collections.Generic;
+
+namespace PxCs.Extensions
```
Comment (JaXt0r) [inline PxCs/Extensions/SoundExtension.cs:6]: This namespace is used for ObjectType extensions only. i.e. IntPtr.MarshalAsString().

How about using another one like Sound/Audio/...?

File: PxCs/Extensions/SoundExtension.cs [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+{
+    internal enum WAVE_FORMAT
+    {
+        UNKNOWN,
+        PCM,
+        ADPCM,
+        IMA_ADPCM = 0x11,
+    }
+
+    internal class Ima
```
Comment (JaXt0r) [inline PxCs/Extensions/SoundExtension.cs:16]: Internal class. Never used it before in PxCs. But I think it's fine here, as we won't expose these classes.

Comment (JaXt0r) [inline PxCs/Extensions/SoundExtension.cs:16]: Can you rename Ima or at least tell in a comment what it means? I don't know this acronym.

**Review by JaXt0r — CHANGES_REQUESTED** [review]
Cool thing! A few updates requested to align more on the other PxCs classes.

And one important thing: Please create a Test for your function(s). You have a lot of examples to adopt on PxCs.Tests. I normally test at least:
1. Is it not executing with a MemoryException etc.
2. Check for some data entries if value is set right. (maybe byte[].Length etc.)

File: PxCs/Data/Sound/PxSoundData.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
+using System;
+using System.Collections.Generic;
+
+namespace PxCs.Data.Sound
+{
+    public class PxSoundData<T>
+    {
+        public Type arrayType;
+        public List<T> sound;
```
Comment (JaXt0r) [inline PxCs/Data/Sound/PxSoundData.cs:9]: Please use an array for performance reasons as the data size is fixed. All other PxCs data uses arrays too. On unity side we leverage dynamic lists.

File: PxCs/Helper/IMAADPCMDecoder.cs [hunk 9]
```diff
@@ -3,7 +3,7 @@
 using System.IO;
 using System.Collections.Generic;
 
-namespace PxCs.Extensions
+namespace PxCs.Helper
```
Comment (JaXt0r) [inline PxCs/Helper/IMAADPCMDecoder.cs:6]: Thanks

File: PxCs/Interface/PxSound.cs
Comment (JaXt0r) [inline PxCs/Interface/PxSound.cs:1]: A test is still missing though.

**Review by JaXt0r — CHANGES_REQUESTED** [review]
Nearly done. Thanks. Only 2-3 smaller recommendations left.

File: PxCs/Extensions/SoundExtension.cs (hunk 7 above)
Comment (JucanAndreiDaniel) [reply PxCs/Extensions/SoundExtension.cs]: Renamed to IMAADPCMDecoder

**Review by JaXt0r — APPROVED** [review]
Lgtm

**Pull request merged**